instrumentation/rdkafka/lib/opentelemetry/instrumentation/rdkafka/patches/consumer.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Rdkafka
+      module Patches
+        # CustomerGetter for needed for propagation
+        class CustomGetter
+          def get(carrier, key)
+            carrier[key.to_sym]
+          end
+
+          def keys(carrier)
+            carrier.keys
+          end
+        end
+        # The Consumer module contains the instrumentation patch for the Consumer class
+        module Consumer
+          def each
+            super do |message|
+              attributes = {
+                'messaging.system' => 'kafka',
+                'messaging.destination' => message.topic,
+                'messaging.destination_kind' => 'topic',
+                'messaging.kafka.partition' => message.partition,
+                'messaging.kafka.offset' => message.offset
+              }
+
+              attributes['messaging.kafka.message_key'] = message.key if message.key
+              parent_context = OpenTelemetry.propagation.extract(message.headers, getter: CustomGetter.new)
+              span_context = OpenTelemetry::Trace.current_span(parent_context).context
+              links = [OpenTelemetry::Trace::Link.new(span_context)] if span_context.valid?
+
+              OpenTelemetry::Context.with_current(parent_context) do
+                tracer.in_span("#{message.topic} process", links: links, attributes: attributes, kind: :consumer) do
+                  yield message
+                end
+              end
+            end
+          end
+
+          def each_batch(max_items: 100, bytes_threshold: Float::INFINITY, timeout_ms: 250, yield_on_error: false, &block)
+            super do |messages, error|
+              if messages.empty?
+                yield messages, error
+              else
+                attributes = {
+                  'messaging.system' => 'kafka',
+                  'messaging.destination_kind' => 'topic',
+                  'messaging.kafka.message_count' => messages.size
+                }
+
+                links = messages.map do |message|
+                  span_context = OpenTelemetry::Trace.current_span(OpenTelemetry.propagation.extract(message.headers, getter: CustomGetter.new)).context
+                  OpenTelemetry::Trace::Link.new(span_context) if span_context.valid?
+                end
+                links.compact!
+
+                tracer.in_span('batch process', attributes: attributes, links: links, kind: :consumer) do |span|
+                  span.record_exception(error) if error
```

Comment:
Do we also want to set span.status in this case? @robertlaurin wdyt? I notice we don't do this in the other kafka instrumentation but it's possible that instrumentation may just be out of date. https://github.com/open-telemetry/opentelemetry-ruby/blob/a8082e6059559719e106d76540c77393293e081e/instrumentation/sidekiq/lib/opentelemetry/instrumentation/sidekiq/middlewares/server/tracer_middleware.rb#L49

Reply:
Good catch, I think it should be this.  IIRC this lib captures errors a little funnily so you have to check for the presence of it instead of it just bubbling up naturally.
```suggestion
                  if error
                    span.record_exception(error)
                    span.status = OpenTelemetry::Trace::Status.error("Unhandled exception of type: #{e.class}") 
                  end
```